File: engines/perl/benchmark.pl
use Time::HiRes qw(gettimeofday);
# use Syntax::Keyword::Try;

sub measure {
  my ($data, $pattern) = @_;
  my $regex = eval { qr/$pattern/ };
  if ($@) {
    die "invalid regex: $@" if $@;
  }

  my $start = Time::HiRes::gettimeofday();
  my $count = () = $data =~ /$regex/g;
  my $elapsed = (Time::HiRes::gettimeofday() - $start) * 1e3;

  printf("%f - %d\n", $elapsed, $count);
}

if (@ARGV != 3) {
  die "Usage: ./benchmark.pl <filename> regex numIterations\n";
}

my ($filename) = @ARGV;

open my $fh, '<', $filename or die 'Could not open file.';
my $text;
read $fh, $data, -s $filename;
close $fh;

my $pattern = $ARGV[1];
my $numIterations = int($ARGV[2]);

for (my $i = 0; $i < $numIterations; $i++) {
  measure($data, $pattern);
}

File: engines/go/benchmark.go
package main

import (
    "bytes"
    "fmt"
    "log"
    "os"
    "regexp"
    "time"
    "strconv"
)

var EXIT_FAILURE int = 1

func measure(data string, pattern string) {

    r, err := regexp.Compile(pattern)
    if err != nil {
        log.Fatal(err)
        os.Exit(EXIT_FAILURE)
    }

    start := time.Now()

    matches := r.FindAllString(data, -1)
    count := len(matches)

    elapsed := time.Since(start)

    fmt.Printf("%f - %v\n", float64(elapsed) / float64(time.Millisecond), count)
}

func main() {
    if len(os.Args) != 4 {
        fmt.Println("Usage: benchmark <filename> regex numIterations")
        os.Exit(1)
    }

    filerc, err := os.Open(os.Args[1])
    if err != nil {
        log.Fatal(err)
    }
    defer filerc.Close()

    buf := new(bytes.Buffer)
    buf.ReadFrom(filerc)
    data := buf.String()

    pattern := os.Args[2]
    numIterations, err := strconv.Atoi(os.Args[3])

    for i := 0; i < numIterations; i++ {
        measure(data, pattern)
    }
}
